credentials.go

```diff
@@ -145,24 +147,73 @@ func ParseKeyFile(location string) ([]byte, error) {
 	return ParseKeyData(data)
 }
 
-var keyDataPattern = regexp.MustCompile(`<Data>(.+)</Data>`)
+type xmlKeyFileData struct {
+	XMLName xml.Name       `xml:"KeyFile"`
+	Meta    xmlKeyFileMeta `xml:"Meta"`
+	Key     xmlKeyFileKey  `xml:"Key"`
+}
+
+type xmlKeyFileMeta struct {
+	Version string `xml:"Version"`
+}
+
+type xmlKeyFileKey struct {
+	Hash string `xml:"Hash,attr"`
+	Data []byte `xml:"Data"`
+}
+
+var whiteSpacePattern = regexp.MustCompile(`\s+`)
+
+const xmlKeyDataHashLength = 4
+
+// var keyDataPattern = regexp.MustCompile(`<Data>(.+)</Data>`)
 
 // ParseKeyData returns the hashed key from a key file in bytes, parsing xml if needed
 func ParseKeyData(data []byte) ([]byte, error) {
-	if keyDataPattern.Match(data) { //If keyfile is in xml form, extract key data
-		base := keyDataPattern.FindSubmatch(data)[1]
-		data = make([]byte, base64.StdEncoding.DecodedLen(len(base)))
-		if _, err := base64.StdEncoding.Decode(data, base); err != nil {
-			return nil, err
+	keyFileData := xmlKeyFileData{}
+	err := xml.Unmarshal(data, &keyFileData)
+	if err == nil {
+		keyFileData.Key.Data = whiteSpacePattern.ReplaceAll(keyFileData.Key.Data, []byte(``))
+
+		switch keyFileData.Meta.Version {
```

Comment:
Probably it is better to define separate decode funcs for every case. This one looks a bit overloaded.

Reply:
Yup, I only threw this together as a first draft

Reply:
I split it out into smaller chunks

I still need to get some test data with different types of keys in order to finish this

Reply:
Do you expect the test data from me? I can prepare and upload it here. Alternatively test data can be generated randomly (with optional xml wrapper), but it that way we can not make sure if it works with original KeePass app.

Reply:
I generated test data by myself, however that is based on my memory of what should work.

I will however not get towards being able to generate database files + corresponding key files in the next 2 days to ensure that that part works

Reply:
Clear. I'll check with manual tests with databases and various key files then. Will report in 2 days as well.